### ISSUE TITLE
Use chrome.storage instead of localStorage

### ISSUE THREAD

**Issue by cowboy** [benjaffe/chrome-okc-plugin#11]
Because then data can be more persistent and synced across browsers!

http://developer.chrome.com/extensions/storage.html

**Comment by sapid** [comment]
Has anyone made any progress on this? I spent the afternoon looking at the branches made by @astronouth7303 and @vectorjohn 6+ months ago. (Incidentally, @astronouth7303, I fixed one of the bugs you had that prevented extension data from binding to and populating the Labels dropdown by delaying the creation of the label menu to document.ready time.) 

I also poked around the chrome.storage.sync docs. Sync's storage limit is 100KB. Right now, each entry in okcp.profileList is 100-150 chars. Since it's all in the ascii range, this translates to 100-150 bytes. At 150 bytes, that's an upper limit of ~600 users stored, which makes me wonder if sync is worth adding, at least using chrome.storage.sync rather than some server with client-side encrypted data. As @astronouth7303 said above, hitting 500 profiles is pretty easy within a year, particularly if you live in a population-dense area.

So, here's some thoughts on the viability of syncing. I think a lot of the stuff in there can be stripped out for the purposes of syncing. Last-visited, Last-modified, even Location - none of this really *needs* to be synced. Location data gets stale over time anyway - people move. 

All that's really necessary for useful sync functionality are the username and associated flags. Everything else can be moved to local storage, and the flags can be compressed a lot - down to a 2 character value, really, by simulating a bitmap with bitwise operators. Even if the number of flags doubles or triples, we're still at 4 characters for the flag value. I think the username character limit is 16?

So, if we use a data structure of the format `Username: flags` our per-profile data consumption is at most `"sixteenCharsHere":31,` or 22 bytes. `102,400/22=4654` which should last a couple of years at least. With settings also synced, we're solidly above 4k users. With some client-side optimizations (for example, keep track of when we last saw people we're keeping data for and keep only the most recent 4500) that's probably workable. In fact, since not all usernames are 16 characters, we're probably looking at a practical limit of ~6k users. I suspect 98% of users won't ever reach that limit... and the last 2% can probably be accommodated by intelligent working-set management.

... Unfortunately, chrome.storage.sync also has a limit of 512 items, which brings us back to square one unless we figure out some kludge-y way of encoding our data into 512 items, none of which are larger than 8192 bytes (372 profiles). Some kind of peculiar hash table might be viable, but unless someone else has already done the work, at that point we're probably better off settling for a working-set of 512 and getting intelligent working-set management working and a smooth way to migrate the entire local database to another client. (e.g. a good backup/restore function that isn't clipboard-based.)

There's also throttling to keep track of, but if we only write to sync every 30 seconds (or 10, or 5) it's not a big problem.

Alternatives might include...
* Google Drive through their API. This would mean elevating the extension's permissions to include chrome.identity, or adding a password prompt and managing tokens. *Downside:* Either manage a lot of files or shunt your entire database each update. While our database is going to be measured in kilobytes, we would still want to write fairly lazily. *Upside:* No central storage of user data; each user gets their own store.
* Google App Engine. Write a server to act as a data store. *Downside:* Write our own auth mechanism, extension depends on the server being up for syncing, and we have to shunt the entire database over the wire since it's an encrypted glob. *Upside:* This might be a little more flexible than Google Drive, and more painless for the user. (Credentials and encryption keys can be generated automatically, stored and synced in chrome.storage.sync. Unfortunately, chrome.storage.sync is not encrypted.)
* WebRTC. This seems like the most attractive option to me. *Upside:* We can massively deduplicate what we send over the wire. (Compare database states, send deltas.) We can probably use chrome.storage.sync to automagically find other open browsers owned by the user. (e.g. use sync to broker connections.) *Downside:* Both clients need to be open at the same time; there's no external store of data. If someone loses their data somehow and doesn't have it stored on another device, it's gone. Probably best coupled with a backup/restore feature.

Compressing flags to a bitmap is probably a worthy goal regardless.

Anyway, the way I see it, the two best options are either 
* using sync and adding fancy working-set management (e.g. sync ~500 most recent profiles viewed, store more stale profiles locally, shuffle as necessary)    
or 
* using WebRTC and adding fancy de-duplication for catching up on what transpired since last sync.

Thoughts?

**Comment by astronouth7303** [comment]
I know in my use cases, having all my instances online at the same time is just not feasible.

Is encrypting user data on the server really necessary?

Maybe looking into CouchDB + PouchDB is worth while? That should make changes and replication super easy. Just need a front-end service to handle user registration.

**Comment by sapid** [comment]
Per #8, Ben said he wasn't interested in a server option due to concerns about data scraping and user privacy, and if he went with one, he would want it to be encrypted. That's why I've been exploring the 

I take back what I said about the database needing to be an encrypted glob. Assuming symmetric encryption, I don't see any reason why we couldn't encrypt before moving things in to the database - we should be able to encrypt each field individually. 

One option would be to store automatically generated keys in chrome.storage.sync. This way, we wouldn't need to handle user registration necessarily. Storing it in chrome.storage.sync would mean that the user would have to authenticate to their browser, which could essentially serve as our auth mechanism. The tradeoff of this sort of seamlessness, of course, is that if a user somehow loses their keys in chrome.storage.sync, they lose their data. 

Traditional user registration would prevent this sort of data loss. It would also let users sign out while they weren't using their computer, or while another user was logged in, which using chrome.storage.sync as a cache using LRU page replacement doesn't offer. (At least, not innately - I suppose there's no reason we couldn't implement that as well.)

Regarding the question of the necessity of encryption, per issue #8, Ben said he wasn't interested in a server option due to concerns about data scraping and user privacy, and if he went with one in the future, he would want it to be encrypted. If all we're storing is username and flags, perhaps that's unnecessary, but people tend to be sensitive about this sort of thing, and I don't think encryption would be very difficult.

Perhaps we could also offer the option to sync or not sync, and make the default not to sync. (That's no change from present circumstances, after all.)

**Comment by astronouth7303** [comment]
Encryption in JS is pretty difficult, simply because libraries aren't wide-spread or well-used. They exist, but I think they're mostly toys and academic. On top of that, the odds of any of them being audited is probably approaching zero.

Even after you have a crypto library, turning crypto primitives into a usable and secure system is hard (this is why standards like OpenPGP and TLS exist). To top it off, key management (esp since loosing the key is catastrophic) is complex without support and user knowledge of it.

In short, rolling your own crypto is hard. 

I guess the real question is, are we protecting from outsiders getting the information or from admins? Protecting from outsiders means other measures can be taken (securing the API interface, encrypting the file system, etc.) without having to roll your own crypto. Protecting from insiders, though. That's a much, much harder nut to crack.

Given what is synced and the probable security of OkCupid (and other sites with this kind of information), I would think per-user security with perfect prevention of insiders is just not worth it. Esp since users already trust developers (to download and install the extension in the first place).

**Comment by sapid** [comment]
I agree that rolling your own crypto is hard, and I'm not advocating that we build a library ourselves; I think that's a dangerous option that is all but guaranteed to do more harm than good. I'd much rather use an implementation like http://www.openpgpjs.org/ or wait for a release from https://github.com/google/end-to-end. 

It's worth noting here, I think, that openpgp.js was audited a little less than a year ago and several critical vulnerabilities were found. https://cure53.de/pentest-report_openpgpjs.pdf I didn't go looking very hard for additional reviews, this shows there's been at least one. I don't know if these vulns have been patched in the past year, but I didn't find issues on the repo through a perhaps somewhat cursory inspection.
